File: TrendCalculator.js
const DecreaseFactorPerSecond = 0.05;

class TrendCalculator
{
	constructor(value = undefined)
	{
		// Current calculated value.
		// @type {Number}
		this._value = value;

		// Highest seen value.
		// @type {Number}
		this._highestValue = value;

		// NOTE: Not used for now since, somehow (idea feliz), we are using
		// this._highestValueAt instead.
		// Last time update() was called (in ms).
		// @type {Number}
		// this._lastUpdateAt = Date.now();

		// Time in which highest value was set.
		// @type {Number}
		this._highestValueAt = Date.now();
	}

	get value()
	{
		return this._value;
	}

	update(value)
	{
		const now = Date.now();
		// const elapsedMs = now - this._lastUpdateAt;
		const elapsedMs = now - this._highestValueAt;
		const elapsedSeconds = elapsedMs / 1000.0;

		// NOTE: Not used for now.
		// this._lastUpdateAt = now;

		// If new value is bigger or equal than current, use it.
		if (value >= this._value)
		{
			this._value = value;
			this._highestValue = value;
			this._highestValueAt = now;
		}
		// Otherwise decrease current value.
		else
		{
			this._value = Math.max(
				value,
				this._highestValue - (this._highestValue * elapsedSeconds * DecreaseFactorPerSecond)
			);
		}
	}
}

module.exports = TrendCalculator;

File: src/TrendCalculator.cpp
#include "TrendCalculator.hpp"
#include <algorithm> // std::min(), std::max()

static constexpr float DecreaseFactorPerSecond{ 0.05f };

TrendCalculator::TrendCalculator()
{

}

uint32_t TrendCalculator::GetValue() const
{
	return this->value;
}

void TrendCalculator::Update(uint32_t value, uint64_t now)
{
	if (this->value == 0u)
	{
		this->value                 = value;
		this->highestValue          = value;
		this->highestValueUpdatedAt = now;

		return;
	}

	// If new value is bigger or equal than current one, use it.
	if (value >= this->value)
	{
		this->value                 = value;
		this->highestValue          = value;
		this->highestValueUpdatedAt = now;
	}
	// Otherwise decrease current value.
	else
	{
		uint64_t elapsedTime = now - this->highestValueUpdatedAt;

		this->value = std::max<uint32_t>(
			value,
			this->highestValue - (this->highestValue * DecreaseFactorPerSecond * (elapsedTime / 1000.0))
		);
	}
}
